Add ZAP Developer Docs

ricekot opened the pull request:
Signed-off-by: ricekot <ricekot@gmail.com>

kingthorin commented on site/content/docs/developer/quick-start-build.md:
````diff
@@ -0,0 +1,146 @@
+## Clone the Repo Locally
+Create a new directory on your system and run the following command in that directory to clone the core ZAP repository.
+```bash
+git clone https://github.com/zaproxy/zaproxy.git
````
Shouldn't people be cloning their fork as the first step? (So that origin properly points to their fork, and they don't end-up with origin/upstream the same).

psiinon replied:
I wanted to to make it as easy as possible for people to get ZAP running from the source code.
Cloning repos complicates things and isnt necessary. But happy to have some text pointing out that if people do plan to contribute back then they should go to step X

ricekot replied:
Well, to prevent that we have a section at the end where they point origin to their fork (using `git remote set-url origin https://github.com/USERNAME/zaproxy.git`) and upstream to the main repo. But if cloning would be easier / more clear I can update the docs. What do you suggest?

kingthorin replied:
@psiinon if these are developer docs isn't the assumption that they're contributing?

Or should we put this under a heading other than "developer"? (`/docs/desktop/run-from-source/`?)

psiinon replied:
Tricky - I want us to have a new section for ZAP internals - this is just the start of that section. But you are right, this page isnt just for ZAP devs. So open to suggestions ...
EG would love to migrate a load of this content: https://github.com/zaproxy/zaproxy/wiki/InternalDetails

kingthorin replied:
Well we all know I over think these things so maybe we leave it as-is since there is a chunk later in the doc that addresses origin/upstream.

If we end-up with lots of questions or devs with problems we can always revisit.

thc202 replied:
Being developer does not imply being a contributor, e.g. you might want to create a custom version of ZAP for yourself. (There are many forks of zaproxy with custom changes that are not contributed back, offline changes should be a higher number I guess.)

kingthorin replied:
Good point. Okay let's leave it as-is for now.

kingthorin commented on site/content/docs/developer/quick-start-build.md:
````diff
@@ -0,0 +1,146 @@
+```
+
+## Running ZAP from Source
+Navigate to the cloned *zaproxy* folder and run the following command (if you are a Windows user, omit the `./`):
````
On windows wouldn't you have to use `gradlew.bat`?

psiinon replied:
I think it will work without, but dont have a windows box to hand to check right now ;)

kingthorin commented on site/content/docs/developer/quick-start-build.md:
```diff
@@ -0,0 +1,146 @@
+
+It is possible that you might be greeted with the following prompt if you already have some other application that is using port 8080.
+
+![ZAP Proxy Port in Use](/img/docs/developer/zap-proxy-port.png)
+
+In this case, simply specify a different port and ZAP should start listening on that port.
```
The one that is being suggested should be free IIRC (it loops 10 times trying to find a free one).

kingthorin commented on site/content/docs/developer/quick-start-build.md:
````diff
@@ -0,0 +1,146 @@
+```
+That’s it! You’re now running ZAP from source! You can confirm this if you see “Dev Build” in the loading window.
+
+![ZAP Dev Build Loading Window](/img/docs/developer/zap-dev-build.png)
````
Might be better to use a green arrow or green box, the blue looks like ZAP blue and part of the UI.

kingthorin commented on site/content/docs/developer/quick-start-build.md:
```diff
@@ -0,0 +1,146 @@
+
+Note that while you have successfully got ZAP running, it doesn’t come with many of the key features that are provided by ZAP add-ons.
+
+## Installing ZAP add-ons
```
```suggestion
## Installing ZAP Add-ons
```

kingthorin commented on site/content/docs/developer/quick-start-build.md:
```diff
@@ -0,0 +1,146 @@
+Note that while you have successfully got ZAP running, it doesn’t come with many of the key features that are provided by ZAP add-ons.
+
+## Installing ZAP add-ons
+As the name suggests, ZAP add-ons are a way to enhance ZAP functionality. There are several add-ons that allow you to do many different things. The *zap-extensions* repository contains the source code for these add-ons. So, the first thing to do is clone *zap-extensions*:
```
```suggestion
As the name suggests, ZAP add-ons are a way to enhance ZAP functionality. There are several add-ons that allow you to do many different things. The *zap-extensions* repository contains the source code for the add-ons maintained by the core team. So, the first thing to do is clone *zap-extensions*:
```

There are third party add-ons maintained elsewhere.

kingthorin commented on site/content/docs/developer/quick-start-build.md:
````diff
@@ -0,0 +1,146 @@
+## Installing ZAP add-ons
+As the name suggests, ZAP add-ons are a way to enhance ZAP functionality. There are several add-ons that allow you to do many different things. The *zap-extensions* repository contains the source code for these add-ons. So, the first thing to do is clone *zap-extensions*:
+```bash
+git clone https://github.com/zaproxy/zap-extensions.git.
````
Same comment re fork/origin/upstream.

psiinon replied:
Ditto ;)

kingthorin commented on site/content/docs/developer/quick-start-build.md:
````diff
@@ -0,0 +1,146 @@
+	│   ├── ascanrules
+```
+
+So, for instance, if you want to install the active and passive scan rules, you would run:
````
```suggestion
So, for instance, if you want to install the Release status Active and Passive scan rules, you would run:
```

kingthorin commented on site/content/docs/developer/quick-start-build.md:
````diff
@@ -0,0 +1,146 @@
+The process of installing the ZAP HUD and the help is identical:
+1. Clone their repositories.
+    ```bash
+    git clone https://github.com/zaproxy/zap-hud.git
+    git clone https://github.com/zaproxy/zap-core-help.git
````
fork/origin/upstream

kingthorin commented on site/content/docs/developer/quick-start-build.md:
```diff
@@ -0,0 +1,146 @@
+
+## For Prospective ZAP Contributors
+If you feel that some part of ZAP can be improved, enhanced or updated in any way, go ahead and open a [pull request](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests)! Here are the steps you need to follow before you can do that:
+1. Clone the main repositories as illustrated above.
+1. Create a [GitHub](https://github.com) account.
+1. Fork the ZAP repositories.
+1. Update the tracked (remote) repositories.
+1. Make changes and submit a PR!
```
I'd suggest swapping or clarifying the fork and clone steps.

kingthorin commented on site/content/docs/developer/quick-start-build.md:
````diff
@@ -0,0 +1,146 @@
+cd ..
+```
+
+### Make changes and submit a PR!
````
```suggestion
### Make Changes and Submit a PR!
```

kingthorin commented on site/content/docs/developer/quick-start-build.md:
```diff
@@ -0,0 +1,146 @@
+### Fork the Repos on GitHub
+Go to the GitHub page of each repository you cloned and [fork it](https://guides.github.com/activities/forking/) to your account (the button is towards the top right of the screen).
+
+![The fork button on GitHub](/img/docs/developer/github-fork.png)
```
Same arrow comment

kingthorin reviewed (changes requested):
This is great

kingthorin commented:
Maybe it's too advanced for this content but I'll mention it anyway: This is the chained one-liner I use for updating all my repos. Doesn't matter what the default branch is because it figures that out for you (master, develop, main, whatever):
```bash
DEFAULT=$(git symbolic-ref --short HEAD) && git fetch -vp origin && git fetch -vp upstream && git merge upstream/$DEFAULT && git push
```

1. It assumes you're in the repo directory and have the default branch checked out.
1. Find the default branch name.
1. Fetch origin verbosely and prune branches that have been removed.
1. Fetch upstream verbosely and prune branches that have been removed.
1. Merge upstream into local.
1. Push back to origin.

psiinon commented:
Great Docs!
Might be worth mentioning the option to install the weekly add-ons.
Also mention that we will be adding more docs relating to makmning contributions in the future?

psiinon commented:
@kingthorin re your power user git command - we should definitely include things like that, but maybe on a future page? I'd really like to keep this one as simple as possible so we dont frighten people off. We can add more pages with more helpful stuff for people to use one they've got ZAP running.

ricekot commented:
Please don't merge this yet, incoming changes...

ricekot commented:
The API key needs to be disabled for `./gradlew installZapAddOn` to work.

ricekot commented:
As discussed on IRC,
- replaced `installZapAddOn` with `copyZapAddOn` (hence eliminating the need for disabling the API key) and
- added a couple of lines for `copyWeeklyAddOns`.

kingthorin reviewed (approved):
LGTM

thc202 commented:
Thanks!
